torch/csrc/autograd/function.h

```diff
@@ -325,6 +339,7 @@ struct TORCH_API Function : std::enable_shared_from_this<Function> {
   std::vector<std::unique_ptr<FunctionPreHook>> pre_hooks_;
   std::vector<std::unique_ptr<FunctionPostHook>> post_hooks_;
```

Comment:
I think the right thing is to keep a copy of the raw pointer for every post-hook in reducer.h, then walk through that list of post hook raw pointers on destruction of the Reducer, and call a new `delete_post_hook` method in `function.h` that removes that specific hook from the array (O(n)). It's a little ugly to register a unique_ptr<> and then keep a raw copy around, but you're *only* using the raw copy as a "key" for deletion, not to actually dereference, so it should be safe. May even want to store the raw pointer as a uintptr_t so it can't be dereferenced.